mysys/crc32/crc32_arm64.c

```diff
@@ -299,3 +300,35 @@ uint32_t crc32c_aarch64(uint32_t crc, const unsigned char *buffer, uint64_t len)
 
 	return (~crc);
 }
+
+/* There are multiple approaches to calculate crc.
+Approach-1: Process 8 bytes then 4 bytes then 2 bytes and then 1 bytes
+Approach-2: Process 8 bytes and remaining workload using 1 bytes
+Apporach-3: Process 64 bytes at once by issuing 8 crc call and remaining
+            using 8/1 combination.
+
+Based on micro-benchmark testing we found that Approach-2 works best especially
+given small chunk of variable data. */
+unsigned int crc32_aarch64(unsigned int crc, const void *buf, size_t len)
+{
+  const uint8_t *buf1= buf;
+  const uint64_t *buf8= (const uint64_t *) (((uintptr_t) buf + 7) & ~7);
+
+  crc= ~crc;
+
+  /* if start pointer is not 8 bytes aligned */
+  while ((buf1 != (const uint8_t *) buf8) && len)
```

Comment:
This loop can be unrolled into Duff's machine:
```
switch ((const uint8_t *) buf8 - buf1)
{
case 7: crc= __crc32b(crc, *buf1++); len--;
case 6: crc= __crc32b(crc, *buf1++); len--;
case 5: crc= __crc32b(crc, *buf1++); len--;
case 4: crc= __crc32b(crc, *buf1++); len--;
case 3: crc= __crc32b(crc, *buf1++); len--;
case 2: crc= __crc32b(crc, *buf1++); len--;
case 1: crc= __crc32b(crc, *buf1++); len--;
case 0: break;
default: __builtin_unreachanble();
}
```

Also, if you compute `uint8_t *end = buf1 + len` you'll be able to write loops like `for (; buf1 != end; ++buf1) {...}` without decrementing `len`.

Reply:
Thanks! I believe that the canonical term is [Duff’s device](https://en.wikipedia.org/wiki/Duff%27s_device).
You seem to have mistyped `__builtin_unreachable()`. That call should be OK here, because this code is depending on GCC or clang anyway.

@mysqlonarm Krunal, could you please test Eugene’s suggestion?

Reply:
I tested the loop in an independent benchmark. The existing loop is slightly faster than the proposed unrolling [existing loop: 13.375, proposed loop: 14.087]. So I think we will leave with the existing loop.